File: examples/scala/ExampleSaxpy.scala
object ExampleSaxpy {
    def main(args: Array[String]) : Unit = {
        //Load Libary
        Executor.loadLibary()

        //Testdata
        val numThreads = 4
        val numBlocks = 4

        val n = numThreads*numBlocks
        val a = 5.1f
        val x = new Array[Float](n)
        val y = new Array[Float](n)
        for (i <- 0 until n) {
            x(i) = i
            y(i) = 2*i
        }

        //Initialize Arguments
        val aArg = FloatArg.createValue(a)
        val xArg = FloatArg.create(x: _*)
        val yArg = FloatArg.create(y: _*)
        val outArg = FloatArg.createOutput(n)
        val nArg = IntArg.createValue(n)

        //Create Program
        val kernelString = Utils.loadFile("kernels/saxpy.cu")
        val saxpy = Program.create(kernelString, "saxpy")

        //Create compiled Kernel
        val saxpyKernel = saxpy.compile()

        //Launch Kernel
        val executionTime = saxpyKernel.launch(numThreads, numBlocks, aArg, xArg, yArg, outArg, nArg)

        //Get Result
        val out = outArg.asFloatArray()

        //Print Result
        println("\nsaxpy-Kernel sucessfully launched:");
        println(executionTime);
        println("\nInput a: " + a);
        println("Input x: [" + x.mkString(", ") + "]");
        println("Input y: [" + y.mkString(", ") + "]");
        println("Result:  [" + out.mkString(", ") + "]");
    }
}

File: examples/java/ExampleSaxpy.java
import java.io.IOException;
import java.util.Arrays;

import yacx.Executor;
import yacx.FloatArg;
import yacx.IntArg;
import yacx.Kernel;
import yacx.KernelArg;
import yacx.KernelTime;
import yacx.Program;
import yacx.Utils;

public class ExampleSaxpy {
    public static void main(String[] args) throws IOException {
        //Load Libary
        Executor.loadLibary();

        //Testdata
        final int numThreads = 4;
        final int numBlocks = 4;

        int n = numThreads*numBlocks;
        float a = 5.1f;
        float[] x = new float[n];
        float[] y = new float[n];
        for (int i = 0; i < n; i++) {
            x[i] = i;
            y[i] = 2*i;
        }

        //Initialize Arguments
        KernelArg aArg, nArg, xArg, yArg;
        FloatArg outArg;
        aArg = FloatArg.createValue(a);
        xArg = FloatArg.create(x);
        yArg = FloatArg.create(y);
        outArg = FloatArg.createOutput(n);
        nArg = IntArg.createValue(n);

        //Create Program
        String kernelString = Utils.loadFile("kernels/saxpy.cu");
        Program saxpy = Program.create(kernelString, "saxpy");

        //Create compiled Kernel
        Kernel saxpyKernel = saxpy.compile();

        //Launch Kernel
        KernelTime executionTime = saxpyKernel.launch(numThreads, numBlocks, aArg, xArg, yArg, outArg, nArg);

        //Get Result
        float[] out = outArg.asFloatArray();

        //Print Result
        System.out.println("\nsaxpy-Kernel sucessfully launched:");
        System.out.println(executionTime);
        System.out.println("\nInput a: " + a);
        System.out.println("Input x: " + Arrays.toString(x));
        System.out.println("Input y: " + Arrays.toString(y));
        System.out.println("Result:  " + Arrays.toString(out));
    }
}